Automattic/wp-calypso
# E2E: reload the page after the log in process completes, but before the test steps take over.

Related to https://github.com/Automattic/wp-calypso/issues/70674.

## Proposed Changes

This PR adds a reload step after the log in process, but before the test steps take over, ostensibly to reset the state before the tests begin as a workaround to the page crash. This workaround was suggested in https://github.com/microsoft/playwright/issues/18137. 

The [original issue](https://github.com/Automattic/wp-calypso/issues/70674.) describes a situation where the teardown code makes an attempt to screenshot the page, but is unable to do so because the page has already crashed. After some investigations, we find that a comment on [microsoft/playwright](https://github.com/microsoft/playwright/issues/18137#issuecomment-1282634092) closely (but not exactly) describes our issue.

During the login process for browsers controlled by Playwright:
- the log in process is initiated, then the form is submitted.
- Calypso authenticates and redirects the page multiple times.
- finally, the page is redirected to the original endpoint they were supposed to land at.
- test steps take over from here and redirect the page to the intended endpoint.

This all occurs with milliseconds to spare because Playwright is able to act on pages several orders of magnitude faster than humans can. The hypothesis is that this is what's causing the eventual page crash.

The proposed idea is to insert a state-resetting `reload` call between the third and fourth steps to "reset" the state. This way, the test steps are not attempt to execute code on a log in process that has not yet fully completed.

## Testing Instructions

Ensure the following build configurations are passing:

I've also ran the proposed changes under a "stress test" of 100 iterations or 20 minutes of execution time, whichever comes first. The idea is to run most of the E2E specs in a single build, repeatedly, in the hopes that a problem either arises (in which case the fix has not worked) or does not arise (in which case the fix has high confidence of working).

For this PR, I ran multiple iterations, first with a limited set of specs and then gradually adding more.

**First iteration** - 15 test steps.
<img width="917" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/7f73d107-55e4-4ef9-a0e6-33de5ebcfdb1">

**Second iteration** - 15 test steps.
<img width="925" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/1a5fddd4-3641-4f37-9d77-aacd337f435c">

**Third iteration** - 48 test steps.
<img width="902" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/29e92de1-3cc4-40a8-ae70-d58245b58fc8">

**Fourth iteration** - 89 test steps.
<img width="905" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/8a6c6ea4-619d-459b-972a-54ce5220df28">

So far, through all iterations the crashing issue has not once cropped up.

## Pre-merge Checklist
